src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ReferenceNumberGeneratorService.java

```diff
@@ -0,0 +1,53 @@
+package uk.gov.companieshouse.api.strikeoffobjections.service.impl;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import uk.gov.companieshouse.api.strikeoffobjections.service.IReferenceNumberGeneratorService;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.util.UUID;
+import java.util.function.Supplier;
+
+@Service
+public class ReferenceNumberGeneratorService implements IReferenceNumberGeneratorService {
+
+    private static final String REFERENCE_NUMBER_PREFIX = "OBJ-";
+
+    private final Supplier<LocalDateTime> dateTimeSupplier;
+
+    @Autowired
+    public ReferenceNumberGeneratorService(Supplier<LocalDateTime> dateTimeSupplier) {
+        this.dateTimeSupplier = dateTimeSupplier;
+    }
+
+    /**
+     * Generates a 12 digit reference number based on the day and a random uuid, and inserts a hyphen every 4 characters
+     * e.g. OBJ-1F3C-A2E4-5D6B
+     * @return a 12 digit reference with hyphens every 4 digits, prefixed with {@value #REFERENCE_NUMBER_PREFIX}
+     */
+    @Override
+    public String generateReferenceNumber() {
+        LocalDate dateNow = dateTimeSupplier.get().toLocalDate();
+        String now = Long.toHexString(dateNow.toEpochDay());
```

Comment:
You get less clashes if you remove the use of the date and just use the UUID. I created a test harness you can use here - https://github.com/companieshouse/strike-off-objections-api/compare/feature/OBJ-162-generate-user-friendly-reference-number...feature/ref-num-test-harness
Over a 5 year period i was getting about half a dozen clashes using the date approach, with UUID it was usually 0 or 1.

Reply:
we could reduce the probability of clashes further by randomising the position of the separators, as long as they are followed by at least 1 char. The ref no. length would stay the same, might have to check with business if this is ok though.

Reply:
Have been doing some tests and it does seem we were trying to be to clever. Just a UUID stripped down to 12 characters seems less likely to cause a collision. I suppose at some point in the future prefixing the ID with the day would make collisions less likely as you know each day can't collide with a previous day. Whereas just a UUID any Id can collide with any id ever. But this seems like it would be greater than 10 years in the future with 5000 a day which is way more than we would ever need. So will just switch to just using UUID.

As for randomising the seperators, I don't think we should do that as the seperators aren't actually part of the id. More just there to break it up for a user reading it over the phone. See bullet point 2 in "Showing reference number to user" https://companieshouse.atlassian.net/browse/OBJ-150